provider/pkg/autonaming/application.go

```diff
@@ -41,29 +59,58 @@ func ApplyAutoNaming(
 // based on its URN name, It ensures the name meets the length constraints, if known.
 // Defaults to the name followed by 7 random hex characters separated by a '-'.
 func getDefaultName(
-	randomSeed []byte,
 	urn resource.URN,
+	engineAutonaming EngineAutonamingConfiguration,
 	autoNamingSpec *metadata.AutoNamingSpec,
 	olds,
 	news resource.PropertyMap,
 	config *AutoNamingConfig,
-) (resource.PropertyValue, error) {
+) (*resource.PropertyValue, bool, error) {
 	sdkName := autoNamingSpec.SdkName
 
 	// Prefer explicitly specified name
 	if v, ok := news[resource.PropertyKey(sdkName)]; ok {
-		return v, nil
+		return &v, true, nil
 	}
 
 	// Fallback to previous name if specified/set.
 	if v, ok := olds[resource.PropertyKey(sdkName)]; ok {
-		return v, nil
+		return &v, true, nil
 	}
 
 	// Generate naming trivia for the resource.
 	namingTriviaApplies, namingTrivia, err := CheckNamingTrivia(sdkName, news, autoNamingSpec.TriviaSpec)
 	if err != nil {
-		return resource.PropertyValue{}, err
+		return nil, false, err
+	}
+
+	if engineAutonaming.AutonamingMode != nil {
+		//panic("yes")
+		switch *engineAutonaming.AutonamingMode {
+		case EngineAutonamingModeDisable:
+			return nil, false, nil
+		case EngineAutonamingModeEnforce:
+			v := resource.NewStringProperty(engineAutonaming.ProposedName)
+			return &v, true, nil
+		case EngineAutonamingModePropose:
+			proposedName := engineAutonaming.ProposedName
+
+			// Apply naming trivia to the generated name.
+			if namingTriviaApplies {
+				proposedName = ApplyTrivia(namingTrivia, proposedName)
+			}
+
+			// Validate the proposed name against the length constraints.
+			if autoNamingSpec.MaxLength > 0 && len(proposedName) > autoNamingSpec.MaxLength {
+				return nil, false, fmt.Errorf("proposed name %q exceeds max length of %d", proposedName, autoNamingSpec.MaxLength)
+			}
+			if autoNamingSpec.MinLength > 0 && len(proposedName) < autoNamingSpec.MinLength {
+				return nil, false, fmt.Errorf("proposed name %q is shorter than min length of %d", proposedName, autoNamingSpec.MinLength)
+			}
+
+			v := resource.NewStringProperty(proposedName)
+			return &v, true, nil
+		}
 	}
 
 	var autoTrim bool
```

Comment:
Is there a way that we can get the new engine autonaming to work with the existing autoNaming config? It seems like it could work with `EngineAutonamingModePropose`. And then maybe we should throw errors if you try to configure it with `Enforce` or `Disable`?

https://github.com/pulumi/pulumi-aws-native/blob/11ac7b3942424e3c4863677b2a6ead0de2ab37bc/provider/pkg/schema/gen.go#L188-L191

Reply:
My original take was a no, but now I think I misunderstood `autoTrim`. It could potentially apply to the Propose mode. `randomSuffixMinLength` won't work because the suffix (or prefix, or whatever format) is already baked into the proposed name.

Reply:
Actually, I'm still not sure about `autoTrim`. Do we have any docs about what its supposed to do exactly? Reading the code, its semantics is roughly "if a random suffix makes the generated name too long, we can trim that random suffix down to N characters, where N is either 1 or whatever the user configured". I don't think there is any way to apply that to the proposed name, given it has an unknown structure. I assume we don't want to be trimming non-random portions of the name.

Reply:
Without/pre`autoTrim` the behavior was to allow the random suffix to be as little as 1 character if that allowed the name to fit within the maxLength requirements.

`autoTrim` introduced a couple of new behaviors.

1. `randomSuffixMinLength` could be used to enforce a minimum length of the random suffix
2. `autoTrim: true`. If the length of the name exceeded the maxLength (including the random suffix) then the name would be trimmed to fix within the maxLength. The part of the name that is trimmed is taken from the middle of the name in order to keep uniqueness. 

It sounds like the `randomSuffixMinLength` might conflict with the new engine functionality. In that case we should throw an error if both are used. Since `autoTrim` removes from the middle it's possible it could work with the new engine behavior, but I think it would also be fine if these two features are mutually exclusive for now (especially if the autotrim behavior could be added to the engine)

Reply:
+1 to what Cory said.
Are we able to make a distinction between whether engine level autonaming is configured for the stack/provider or only for a certain resource.

I feel like `autoTrim` + stack/provider level autonaming config should be an invalid case, but `autoTrim` + resource level autonaming config should be valid because it is essentially an escape hatch.

Reply:
> Are we able to make a distinction between whether engine level autonaming is configured for the stack/provider or only for a certain resource?

No

> I feel like autoTrim + stack/provider level autonaming config should be an invalid case, but autoTrim + resource level autonaming config should be valid because it is essentially an escape hatch.

Could you please explain why this distinction? If a user configures a pattern like `${stack}-${name}-${hex(6)}` on the provider level for all AWS Native resources vs. does so for a give `aws-native:foo:Bar` - how does that change the expectation of the user of whether some middle part of the generated name will be trimmed or not?

In any case, the question is a bit hypothetical because the provider can't tell the difference with the current design. I think I'm inclined to do the following for `Propose` mode:

1. If `RandomSuffixMinLength` is configured, throw an error, we can't do anything with it.
2. If `AutoTrim` is true and `autoNamingSpec.MaxLength` is defined, call `trimName(ProposedName, autoNamingSpec.MaxLength)` and hope it does what the user wanted.

But we can also keep it simple and throw an error for now.

Thoughts?

Reply:
I think it's fine for now to just say that you can't configure both the engine autonaming _and_ the provider autoName config and throw an error/warning if you have explicitly configured both.

I can create an issue for supporting a `autoTrim` type feature in the engine, would that be in pu/pu?

Reply:
Added returning an error if both engine and provider configs are specified. Also, renamed a few types and variables to make it consistent and more clear.

> I can create an issue for supporting a autoTrim type feature in the engine, would that be in pu/pu?

yes